ipapython/ipa_log_manager.py

```diff
@@ -141,12 +190,37 @@ def filter(self, record):
                 record.levelno >= self.level)
 
 
+class InfoCommandFilter(logging.Filter):
+    """
+    The InfoCommandFilter filters out info_cmd log messages for the log file.
+    info_cmd log messages should only be visble on the command line.
+    """
+
+    def filter(self, record):
+        return not getattr(record, "cmd_output", False)
+
+
 class Formatter(logging.Formatter):
     def __init__(
             self, fmt=LOGGING_FORMAT_STDOUT, datefmt=ISO8601_UTC_DATETIME_FMT):
         super(Formatter, self).__init__(fmt, datefmt)
         self.converter = time.gmtime
+        self.cmd_output_formatter = logging.Formatter(
+            LOGGING_FORMAT_CMD_OUTPUT)
+
+    def format(self, record):
+        """
+        Format the specified record as text.
+
+        Uses LOGGING_FORMAT_CMD_OUTPUT for info_cmd, otherwise normal logging
+        format. For info_cmd record.__dict__["cmd_output"] is set to True.
+        """
+        if record.levelno >= logging.INFO and \
```

Comment:
This logic should not be in the formatter but rather in the handler's formatter configuration. With this approach the debug logs will miss context and time information for ``logcm`` events. Please use a different formatter for log files and stdout logging here:

https://github.com/freeipa/freeipa/blob/b79ea6a6e92d224859c2d4533b99b3c15d3711ca/ipapython/ipa_log_manager.py#L166-L179

https://github.com/freeipa/freeipa/blob/b79ea6a6e92d224859c2d4533b99b3c15d3711ca/ipalib/plugable.py#L450-L522

Reply:
This is the wrong approach, see [flow diagram](https://docs.python.org/3/howto/logging.html#logging-flow). The logic should be in the handler configuration. You have to use multiple handlers:

1. One stdout handler with a filter for ``cmd_output`` records, a filter for ``level < WARNING`` records, and a ``cmd_output_format`` formatter.
2. One stderr handler with a filter for ``cmd_output`` records, a filter for ``level >= WARNING`` records, and a ``cmd_output_format`` formatter.
3. One stdout handler with a filter for non-``cmd_output`` records and ``console_format`` formatter.

Handler 3 could be split into two separate handlers. I think it would be a good idea to not print a full traceback in ``cmd_output_formatter`` (see ``formatException``).